ESPHome fails to compile for Raspberry Pi Pico W

### The problem

ESPHome 2024.9.1 fails to compile for Raspberry Pi Pico W.
Running HomeAssistant Core 2024.9.3, Supervisor 2024.09.1, Operating System 13.1, Frontend 20240909.1 on a Raspberry Pi 3b.
The Pico W works fine on 2024.8.5 firmware but will not update.

There seems to be other reports of issues with the recent updates to ESPHome here:
[https://community.home-assistant.io/t/esphome-problems-building-for-rp2040-rpipicow/777019](url)

```
INFO ESPHome 2024.9.1
INFO Reading configuration /config/esphome/lounge-subwoofer-pico.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing lounge-subwoofer-pico (board: rpipicow; framework: arduino; platform: https://github.com/maxgerhardt/platform-raspberrypi.git)
--------------------------------------------------------------------------------
HARDWARE: RP2040 133MHz, 256KB RAM, 2MB Flash
 - framework-arduinopico @ 1.30904.0 (3.9.4) 
 - tool-picotool-rp2040-earlephilhower @ 5.140200.240929 (14.2.0) 
Flash size: 2.00MB
Sketch size: 1.00MB
Filesystem size: 1.00MB
Maximium Sketch size: 1044480 EEPROM start: 0x101ff000 Filesystem start: 0x100ff000 Filesystem end: 0x101ff000
Dependency Graph
|-- WiFi @ 1.0.0
|-- LEAmDNS @ 1.2
|-- Updater @ 1.0
|-- noise-c @ 0.1.6
|-- MD5Builder @ 1.0.0
|-- lwIP-Ethernet @ 1
|-- lwIP_CYW43 @ 1
|-- SPI @ 1.0
Linking .pioenvs/lounge-subwoofer-pico/firmware.elf
/data/cache/platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/14.2.0/../../../../arm-none-eabi/bin/ld: /data/cache/platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/14.2.0/../../../../arm-none-eabi/lib/thumb/libc.a(libc_a-findfp.o): in function `std':
/workdir/repo/newlib/newlib/libc/stdio/findfp.c:93:(.text+0xda): undefined reference to `__retarget_lock_init_recursive'
/data/cache/platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/14.2.0/../../../../arm-none-eabi/bin/ld: /workdir/repo/newlib/newlib/libc/stdio/findfp.c:93:(.text+0x124): undefined reference to `__retarget_lock_init_recursive'
/data/cache/platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/14.2.0/../../../../arm-none-eabi/bin/ld: /workdir/repo/newlib/newlib/libc/stdio/findfp.c:93:(.text+0x17c): undefined reference to `__retarget_lock_init_recursive'
/data/cache/platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/14.2.0/../../../../arm-none-eabi/bin/ld: /data/cache/platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/14.2.0/../../../../arm-none-eabi/lib/thumb/libc.a(libc_a-findfp.o): in function `__sfp':
/workdir/repo/newlib/newlib/libc/stdio/findfp.c:201:(.text+0x1ea): undefined reference to `__retarget_lock_init_recursive'
collect2: error: ld returned 1 exit status
*** [.pioenvs/lounge-subwoofer-pico/firmware.elf] Error 1
========================= [FAILED] Took 44.73 seconds =========================
```

### Which version of ESPHome has the issue?

2024.9.1

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2024.9.3

### What platform are you using?

RP2040

### Board

Raspberry Pi Pico W

### Component causing the issue

_No response_

### Example YAML snippet

```yaml
esphome:
  name: lounge-subwoofer-pico
  friendly_name: Lounge Subwoofer Pico

rp2040:
  board: rpipicow
  framework:
    # Required until https://github.com/platformio/platform-raspberrypi/pull/36 is merged
    platform_version: https://github.com/maxgerhardt/platform-raspberrypi.git

# Enable logging
logger:

# Enable Home Assistant API
api:
  encryption:
    key: "cQSsoMBtPdQ2XYpWEWp7+2YqIwvBkV1/O1W7jCZ5f9E="

ota:
  - platform: esphome
    password: "e35bb9b6cf370418167faf19e9552dd9"

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

  # Enable fallback hotspot in case wifi connection fails
  ap:
    ssid: "Lounge-Subwoofer-Pico"
    password: "Qr0Ix36qGneY"
    
# Switch component for subwoofer control
switch:
  - platform: gpio
    name: Subwoofer
    pin: 16
    inverted: true  # This line inverts the logic
  # Optional: Set an icon for the switch in HA
    icon: mdi:subwoofer
    restore_mode: ALWAYS_OFF  # Ensures relay starts off on boot
```


### Anything in the logs that might be useful for us?

```txt
INFO ESPHome 2024.9.1
INFO Reading configuration /config/esphome/lounge-subwoofer-pico.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing lounge-subwoofer-pico (board: rpipicow; framework: arduino; platform: https://github.com/maxgerhardt/platform-raspberrypi.git)
--------------------------------------------------------------------------------
HARDWARE: RP2040 133MHz, 256KB RAM, 2MB Flash
 - framework-arduinopico @ 1.30904.0 (3.9.4) 
 - tool-picotool-rp2040-earlephilhower @ 5.140200.240929 (14.2.0) 
Flash size: 2.00MB
Sketch size: 1.00MB
Filesystem size: 1.00MB
Maximium Sketch size: 1044480 EEPROM start: 0x101ff000 Filesystem start: 0x100ff000 Filesystem end: 0x101ff000
Dependency Graph
|-- WiFi @ 1.0.0
|-- LEAmDNS @ 1.2
|-- Updater @ 1.0
|-- noise-c @ 0.1.6
|-- MD5Builder @ 1.0.0
|-- lwIP-Ethernet @ 1
|-- lwIP_CYW43 @ 1
|-- SPI @ 1.0
Linking .pioenvs/lounge-subwoofer-pico/firmware.elf
/data/cache/platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/14.2.0/../../../../arm-none-eabi/bin/ld: /data/cache/platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/14.2.0/../../../../arm-none-eabi/lib/thumb/libc.a(libc_a-findfp.o): in function `std':
/workdir/repo/newlib/newlib/libc/stdio/findfp.c:93:(.text+0xda): undefined reference to `__retarget_lock_init_recursive'
/data/cache/platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/14.2.0/../../../../arm-none-eabi/bin/ld: /workdir/repo/newlib/newlib/libc/stdio/findfp.c:93:(.text+0x124): undefined reference to `__retarget_lock_init_recursive'
/data/cache/platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/14.2.0/../../../../arm-none-eabi/bin/ld: /workdir/repo/newlib/newlib/libc/stdio/findfp.c:93:(.text+0x17c): undefined reference to `__retarget_lock_init_recursive'
/data/cache/platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/14.2.0/../../../../arm-none-eabi/bin/ld: /data/cache/platformio/packages/toolchain-rp2040-earlephilhower/bin/../lib/gcc/arm-none-eabi/14.2.0/../../../../arm-none-eabi/lib/thumb/libc.a(libc_a-findfp.o): in function `__sfp':
/workdir/repo/newlib/newlib/libc/stdio/findfp.c:201:(.text+0x1ea): undefined reference to `__retarget_lock_init_recursive'
collect2: error: ld returned 1 exit status
*** [.pioenvs/lounge-subwoofer-pico/firmware.elf] Error 1
========================= [FAILED] Took 44.73 seconds =========================
```


### Additional information

_No response_

--- a/esphome/components/rp2040/__init__.py
+++ b/esphome/components/rp2040/__init__.py
@@ -71,6 +71,14 @@ def _format_framework_arduino_version(ver: cv.Version) -> str:
     # return f"~1.{ver.major}{ver.minor:02d}{ver.patch:02d}.0"
 
 
+def _parse_platform_version(value):
+    value = cv.string(value)
+    if value.startswith("http"):
+        return value
+
+    return f"https://github.com/maxgerhardt/platform-raspberrypi.git#{value}"
+
+
 # NOTE: Keep this in mind when updating the recommended version:
 #  * The new version needs to be thoroughly validated before changing the
 #    recommended version as otherwise a bunch of devices could be bricked
@@ -82,10 +90,9 @@ def _format_framework_arduino_version(ver: cv.Version) -> str:
 #  - https://api.registry.platformio.org/v3/packages/earlephilhower/tool/framework-arduinopico
 RECOMMENDED_ARDUINO_FRAMEWORK_VERSION = cv.Version(3, 9, 4)
 
-# The platformio/raspberrypi version to use for arduino frameworks
-#  - https://github.com/platformio/platform-raspberrypi/releases
-#  - https://api.registry.platformio.org/v3/packages/platformio/platform/raspberrypi
-ARDUINO_PLATFORM_VERSION = cv.Version(1, 13, 0)
+# The raspberrypi platform version to use for arduino frameworks
+#  - https://github.com/maxgerhardt/platform-raspberrypi/tags
+RECOMMENDED_ARDUINO_PLATFORM_VERSION = "v1.2.0-gcc12"
 
 
 def _arduino_check_versions(value):
@@ -111,7 +118,8 @@ def _arduino_check_versions(value):
     value[CONF_SOURCE] = source or _format_framework_arduino_version(version)
 
     value[CONF_PLATFORM_VERSION] = value.get(
-        CONF_PLATFORM_VERSION, _parse_platform_version(str(ARDUINO_PLATFORM_VERSION))
+        CONF_PLATFORM_VERSION,
+        _parse_platform_version(RECOMMENDED_ARDUINO_PLATFORM_VERSION),
     )
 
     if version != RECOMMENDED_ARDUINO_FRAMEWORK_VERSION:
@@ -120,15 +128,6 @@ def _arduino_check_versions(value):
         )
 
     return value
-
-
-def _parse_platform_version(value):
-    try:
-        # if platform version is a valid version constraint, prefix the default package
-        cv.platformio_version_constraint(value)
-        return f"platformio/raspberrypi@{value}"
-    except cv.Invalid:
-        return value
 
 
 ARDUINO_FRAMEWORK_SCHEMA = cv.All(

--- a/esphome/wizard.py
+++ b/esphome/wizard.py
@@ -90,9 +90,6 @@ esp32:
 RP2040_CONFIG = """
 rp2040:
   board: {board}
-  framework:
-    # Required until https://github.com/platformio/platform-raspberrypi/pull/36 is merged
-    platform_version: https://github.com/maxgerhardt/platform-raspberrypi.git#5e87ae34ca025274df25b3303e9e9cb6c120123c
 """
 
 BK72XX_CONFIG = """
@@ -184,18 +181,14 @@ def wizard_file(**kwargs):
     password: "{fallback_psk}"
 
 captive_portal:
-    """.format(
-            **kwargs
-        )
+    """.format(**kwargs)
     else:
         config += """
   # Enable fallback hotspot in case wifi connection fails
   ap:
     ssid: "{fallback_name}"
     password: "{fallback_psk}"
-    """.format(
-            **kwargs
-        )
+    """.format(**kwargs)
 
     return config
 

--- a/esphome/wizard.py
+++ b/esphome/wizard.py
@@ -181,14 +181,18 @@ def wizard_file(**kwargs):
     password: "{fallback_psk}"
 
 captive_portal:
-    """.format(**kwargs)
+    """.format(
+            **kwargs
+        )
     else:
         config += """
   # Enable fallback hotspot in case wifi connection fails
   ap:
     ssid: "{fallback_name}"
     password: "{fallback_psk}"
-    """.format(**kwargs)
+    """.format(
+            **kwargs
+        )
 
     return config
 

--- a/platformio.ini
+++ b/platformio.ini
@@ -165,7 +165,7 @@ platform_packages =
 extends = common:arduino
 board_build.filesystem_size = 0.5m
 
-platform = https://github.com/maxgerhardt/platform-raspberrypi.git#5e87ae34ca025274df25b3303e9e9cb6c120123c
+platform = https://github.com/maxgerhardt/platform-raspberrypi.git#v1.2.0-gcc12
 platform_packages =
     ; earlephilhower/framework-arduinopico@~1.20602.0 ; Cannot use the platformio package until old releases stop getting deleted
     earlephilhower/framework-arduinopico@https://github.com/earlephilhower/arduino-pico/releases/download/3.9.4/rp2040-3.9.4.zip

--- a/tests/test_build_components/build_components_base.rp2040-ard.yaml
+++ b/tests/test_build_components/build_components_base.rp2040-ard.yaml
@@ -4,9 +4,6 @@ esphome:
 
 rp2040:
   board: rpipicow
-  framework:
-    # Waiting for https://github.com/platformio/platform-raspberrypi/pull/36
-    platform_version: https://github.com/maxgerhardt/platform-raspberrypi.git#5e87ae34ca025274df25b3303e9e9cb6c120123c
 
 logger:
   level: VERY_VERBOSE

